.github/workflows/release.yml

```diff
@@ -199,64 +259,65 @@ jobs:
           path: ./conda_packages
           merge-multiple: true
 
-      - name: Get Version Number
-        if: contains(github.ref, '/tags')
-        run: |
-          VERSION="${GITHUB_REF/refs\/tags\/v/}"
-          echo "VERSION=$VERSION" >> $GITHUB_ENV
-
       - name: Create GitHub Release and Upload Release Asset
-        if: contains(github.ref, '/tags')
-        uses: softprops/action-gh-release@v1
+        if: env.RELEASE == 'true'
+        uses: softprops/action-gh-release@v2
         env:
           GITHUB_TOKEN: ${{ secrets.GITHUB_TOKEN }}
         with:
           tag_name: ${{ github.event.ref }}
-          name: ${{ env.PACKAGE_NAME }} ${{ env.VERSION }}
+          name: ${{ env.PKGNAME }} ${{ env.PKGVER }}
           draft: true
           prerelease: ${{ contains(github.event.ref, 'rc') }}
-          files: ./dist/${{ env.PACKAGE_NAME }}-${{ env.VERSION }}.tar.gz
+          files: ./dist/${{ env.PKGNAME }}-${{ env.PKGVER }}.tar.gz
 
       - name: Publish on TestPyPI
+        if: env.TESTPYPI_UPLOAD == 'true'
         uses: pypa/gh-action-pypi-publish@release/v1
         with:
           user: __token__
-          password: ${{ secrets.testpypi }}
+          password: ${{ secrets.TESTPYPI }}
           repository-url: https://test.pypi.org/legacy/
 
       - name: Publish on PyPI
-        if: contains(github.ref, '/tags')
+        if: env.PYPI_UPLOAD == 'true'
         uses: pypa/gh-action-pypi-publish@release/v1
         with:
           user: __token__
-          password: ${{ secrets.pypi }}
+          password: ${{ secrets.PYPI }}
```

Comment:
Question and a request.

1) What is the appetite for switching to [trusted publisher](https://packaging.python.org/en/latest/guides/publishing-package-distribution-releases-using-github-actions-ci-cd-workflows/) releasing (instead of API keys) for PyPI? In the spirit of simplifying the workflow, it is actually quite nice since API keys don't need to be configured. The only github thing that _should_ be done is configuring a named environment, and that is optional.

2) Strongly related to the above, could I request that the release job be broken up into separate jobs? I recently got bit by this workflow. Using trusted publisher releasing on testPyPI and PyPI combined with PEP 740 just getting supported causes annoying breakage. Problem is attestation files are built locally on upload and pollute the dist directory. If you release to testPyPI and PyPI in the same job, the files collide and break whichever one is second.

It has apparently been the recommendation of PyPA to use separate jobs (and named environments) for all release jobs for a while now, and they are [very adamant](https://github.com/pypa/gh-action-pypi-publish/issues/283#issuecomment-2499302876) that using a single job for multiple releases is Not Supported (TM).

Reply:
I've split the releases into multiple jobs - for consistency have split them all up.

I'm out of the loop with new publishing developments (have been vaguely ignoring prods and notifications I've seen pop up in my peripheral vision about it), but I'm keen to do whatever will delay bit-rot the longest and minimise the need to configure individual repositories (though it sounds like you need to do manual configuration on PyPI and Test PyPI per project still, oh well). But yeah if the existing method is deprecated it sounds like we should move to this trusted publisher release mechanism.

I don't know what attestations are, but are these a separate thing just happening behind the scenes now regardless, and we just need to split the jobs up to not mess with it?

Then there's digital signing of the GitHub release tarballs as well - that's documented in the trusted publisher releasing documentation, that sounds optional, and I'm not sure how it ties in with the other things or is totally unrelated.

> The only github thing that should be done is configuring a named environment, and that is optional.

Cool, so you're saying you could create the environment in the GitHub UI, or you could just do the following in the workflow as per the example in the docs?

```yaml
  publish-to-pypi:
    name: >-
      Publish Python 🐍 distribution 📦 to PyPI
    if: startsWith(github.ref, 'refs/tags/')  # only publish to PyPI on tag pushes
    needs:
    - build
    runs-on: ubuntu-latest
    environment:
      name: pypi
      url: https://pypi.org/p/<package-name>  # Replace <package-name> with your PyPI project name
    permissions:
      id-token: write  # IMPORTANT: mandatory for trusted publishing
```

If so that's fine - we'd insert `<package-name>` programmatically there from the introspected PKGNAME variable rather than modifying the workflow file per project (which I am EXTREMELY keen to avoid).


Let me know if I've misunderstood anything. Happy to give this a go.

Reply:
> Cool, so you're saying you could create the environment in the GitHub UI, or you could just do the following in the workflow as per the example in the docs?

I have been creating them through the UI, but reading some docs I think they are automatically created when the workflow is first run. Which is nice. One less manual thing to keep track of.

Overall, I've been fairly happy with the trusted publisher stuff. It is a little annoying to have to set it up for both PyPI and TestPyPI, but it's all on a single screen and all the inputs will basically be identical for each repo.

Also, I am super looking forward to the refreshed workflow. Keeping them functional has become a huge pain.

Reply:
> Cool, so you're saying you could create the environment in the GitHub UI, or you could just do the following in the workflow as per the example in the docs?

The environment is auto-created. But it's created without any useful configuration (like having scope secrets). One of my favorite uses, though, is the required reviewers feature — if you enable it, it's like adding a pause in your workflow run — it'll hold and only start the job with the environment once a human clicked a button. This is useful in that it gives one more control over the final pre-release check/gate and gives a chance to download the workflow-stored artifact and inspect them as much as you like.

If you don't configure this, the OIDC-based mechanism in Trusted Publishing will still use the env name to match the expectations you've set up on the PyPI side (the name should be the same).


> If so that's fine - we'd insert `<package-name>` programmatically there from the introspected PKGNAME variable rather than modifying the workflow file per project (which I am EXTREMELY keen to avoid).

Moreover, you can also compute the version-specific page. These links will be preserved on the deployments page. Example: https://github.com/cherrypy/cheroot/deployments/pypi. URL rendering: https://github.com/cherrypy/cheroot/actions/runs/8786461123/workflow#L1181.